### PR TITLE
[Feature] Support training on a single channel image dataset.

### PR DESCRIPTION
**Pull request by landhill**
Support training on a single channel image dataset in `project/`

### CODE REVIEW

File: projects/single_channel/single_channel.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+for file in file_list:
+    if imghdr.what(path + '/' + file) != 'jpeg':
+        continue
+    o_img = Image.open(path + '/' + file)
+    L_img = o_img.convert('L')
+    L_img.save(save_path + '/' + file)
+
+path = 'projects/single_channel/data/balloon/val'
+save_path = 'projects/single_channel/data/balloon/val'
+file_list = os.listdir(path)
```
Comment (PeterH0323) [inline projects/single_channel/single_channel.py:21]: 这几个路径变量和上面的路径变量请用命令行入参代替，参考：

https://github.com/open-mmlab/mmyolo/blob/b0cf0732412d9d7a69f5b7fb9368dab8be6983cc/demo/image_demo.py#L16-L44

File: projects/single_channel/single_channel.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+# Grayscale conversion of each image
+for file in file_list:
```
Comment (PeterH0323) [inline projects/single_channel/single_channel.py:24]: 函数总入口请使用 main 包裹，参考

https://github.com/open-mmlab/mmyolo/blob/b0cf0732412d9d7a69f5b7fb9368dab8be6983cc/tools/train.py#L115-L116

File: projects/single_channel/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# MMYOLO Model #409
```
Comment (PeterH0323) [inline projects/single_channel/README.md:1]:
```suggestion
# Training on a single channel image dataset
```

File: projects/single_channel/README.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+```shell
+python tools/misc/download_dataset.py --dataset-name balloon --save-dir projects/single_channel/data --unzip
+python projects/single_channel/balloon2coco_single_channel.py
+python projects/single_channel/single_channel.py
+```
+
+### Training
+
+In the configuration file, `_base_` and `data_root` need to be modified to the corresponding dataset paths.
```
Comment (PeterH0323) [inline projects/single_channel/README.md:23]: 请加上数据集的最终文件结构示意图

**Comment by landhill** [comment]
已作修改，不确定是否满足要求

File: projects/single_channel/README.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
+# Training on a single channel image dataset
```
Comment (hhaAndroid) [inline projects/single_channel/README.md:1]: 这个暂时不需要用 project 形式支持。从目前来看功能比较简单。应该采用如下做法：

1. 在 docs/zh_cn/advanced_guides/how_to.md 文档里面新增一个部分，描述 Training on a single channel image dataset
2. 在 这个文档中以 configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py 为base 配置，直接告诉大家如何新增 yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_channel.py文件，不需要真的上传一个 py 文件
3. 其中由于这个数据集是三通道的，你就假装插入一个 3通道转 1通道的pipeline 即可，目的是让大家可以按照你的步骤跑通就行
这个 pr ，最终就只是修改了 how-to.md 而已，不需要上传其他东西。

File: docs/zh_cn/advanced_guides/how_to.md [hunk 6]
```diff
@@ -552,3 +552,125 @@ python ./tools/train.py \
 - `randomness.diff_rank_seed=True`，根据 rank 来设置不同的种子，`diff_rank_seed` 默认为 False。
 
 - `randomness.deterministic=True`，把 cuDNN 后端确定性选项设置为 True，即把`torch.backends.cudnn.deterministic` 设为 True，把 `torch.backends.cudnn.benchmark` 设为False。`deterministic` 默认为 False。更多细节见 https://pytorch.org/docs/stable/notes/randomness.html。
+
+## 在一个单通道图像数据集上进行训练
+
+### 数据集预处理
+
+以ballon数据集为例,如果你使用的是自定义灰度图像数据集，你可以跳过这一步。自定义数据集的处理训练可参照[自定义数据集 标注+训练+测试+部署 全流程](../user_guides/custom_dataset.md)。
```
Comment (PeterH0323) [inline docs/zh_cn/advanced_guides/how_to.md:560]:
```suggestion
以 ballon 数据集为例，如果你使用的是自定义灰度图像数据集，你可以跳过这一步。

自定义数据集的处理训练可参照[自定义数据集 标注+训练+测试+部署 全流程](../user_guides/custom_dataset.md)。
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 7]
```diff
@@ -552,3 +552,125 @@ python ./tools/train.py \
...
+
+运行下面的命令，将数据集图片替换为单通道图片。
+
+```shell
+python projects/single_channel/single_channel.py --path projects/single_channel/data/balloon
+```
+
+图像单通道转换示例代码:
+
+```shell
```
Comment (PeterH0323) [inline docs/zh_cn/advanced_guides/how_to.md:576]:
```suggestion
```python
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 8]
```diff
@@ -552,3 +552,125 @@ python ./tools/train.py \
...
+        L_img.save(save_path + '/' + file)
+
+
+if (__name__ == '__main__'):
+    main()
+```
+
+### 新建配置文件并训练
+
+以`configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py`为`base`配置，新增 `yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_channel.py`文件。
```
Comment (PeterH0323) [inline docs/zh_cn/advanced_guides/how_to.md:628]:
```suggestion
以 `configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py` 为 `base` 配置，新增 `yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_channel.py` 文件。
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 9]
```diff
@@ -552,3 +552,125 @@ python ./tools/train.py \
...
+
+```shell
```
Comment (PeterH0323) [inline docs/zh_cn/advanced_guides/how_to.md:630]:
```suggestion
```python
```

File: docs/en/advanced_guides/how_to.md [hunk 10]
```diff
@@ -546,3 +546,125 @@ python ./tools/train.py \
 - `randomness.seed=2023`, set the random seed to 2023.
 - `randomness.diff_rank_seed=True`, set different seeds according to global rank. Defaults to False.
 - `randomness.deterministic=True`, set the deterministic option for cuDNN backend, i.e., set `torch.backends.cudnn.deterministic` to True and `torch.backends.cudnn.benchmark` to False. Defaults to False. See https://pytorch.org/docs/stable/notes/randomness.html for more details.
+
+## Training on a single channel image dataset
+
+### Data set preparation
+
+Take the ballon dataset as an example, if you are using a custom grayscale image dataset, you can skip this step. The processing training of the custom dataset can be found in[Annotation-to-deployment workflow for custom dataset](../user_guides/custom_dataset.md)。
```
Comment (PeterH0323) [inline docs/en/advanced_guides/how_to.md:554]:
```suggestion
Take the ballon dataset as an example, if you are using a custom grayscale image dataset, you can skip this step.

The processing training of the custom dataset can be found in[Annotation-to-deployment workflow for custom dataset](../user_guides/custom_dataset.md)。
```

File: docs/en/advanced_guides/how_to.md [hunk 11]
```diff
@@ -546,3 +546,125 @@ python ./tools/train.py \
...
+
+Run the following command to replace the dataset image with a single channel image.
+
+```shell
+python projects/single_channel/single_channel.py --path projects/single_channel/data/balloon
+```
+
+Image single channel conversion sample code:
+
+```shell
```
Comment (PeterH0323) [inline docs/en/advanced_guides/how_to.md:570]:
```suggestion
```python
```

File: docs/en/advanced_guides/how_to.md [hunk 12]
```diff
@@ -546,3 +546,125 @@ python ./tools/train.py \
...
+
+if (__name__ == '__main__'):
+    main()
+```
+
+### Create a new profile and train
+
+Take `configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py` as `base` configuration, add `yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_ channel.py` file.
+
+```shell
```
Comment (PeterH0323) [inline docs/en/advanced_guides/how_to.md:624]:
```suggestion
```python
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 13]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+python tools/misc/download_dataset.py --dataset-name balloon --save-dir projects/single_channel/data --unzip
+python projects/single_channel/balloon2coco_single_channel.py
+#--save-dir 示例数据集存储路径
+```
+
+运行下面的命令，将数据集图片替换为单通道图片。
+
+```shell
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply docs/zh_cn/advanced_guides/how_to.md:572]: 1. 代码如下所示
```python 
...
```
2. 运行
将上述脚本命名为 `cvt_single_channel.py`,  运行命令为：
```shell
python cvt_single_channel.py  --path projects/single_channel/data/balloon
```
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 14]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+
+from PIL import Image
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='data_path')
+    parser.add_argument(
+        '--path',
+        type=str,
+        default='projects/single_channel/data/balloon',
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:592]: 不要设置 default

File: docs/zh_cn/advanced_guides/how_to.md [hunk 15]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+        L_img.save(save_path + '/' + file)
+
+
+if (__name__ == '__main__'):
+    main()
+```
+
+### 新建配置文件并训练
+
+以 `configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py`为 `base` 配置，新增 `yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_channel.py` 文件。
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:630]:
```suggestion
以 `configs/yolov5/yolov5_s-v61_syncbn_fast_1xb4-300e_balloon.py`为 `base` 配置，在同级配置路径下新增 `yolov5_s-v61_syncbn_fast_1xb4-300e_balloon_single_channel.py` 文件。
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 16]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+
+```python
+_base_ = '../../../configs/yolov5/yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py'
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:633]:
```suggestion
_base_ = 'yolov5_s-v61_syncbn_fast_8xb16-300e_coco.py'
```

File: docs/zh_cn/advanced_guides/how_to.md [hunk 17]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+
+data_root = '../../../projects/single_channel/data/balloon/'
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:635]: 不用改

File: docs/zh_cn/advanced_guides/how_to.md [hunk 18]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+
+train_batch_size_per_gpu = 4
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:637]: 有些地方如果没有改，就不用重写的。继承的配置只需要写你修改的地方就行。

File: docs/zh_cn/advanced_guides/how_to.md [hunk 19]
```diff
@@ -552,3 +552,127 @@ python ./tools/train.py \
...
+val_evaluator = dict(ann_file=data_root + 'val.json')
+
+test_evaluator = val_evaluator
+
+model = dict(bbox_head=dict(head_module=dict(num_classes=1)))
+
+default_hooks = dict(logger=dict(interval=1))
+```
+
+训练测试结果：
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:674]: 要保证这个配置是直接可以跑的，麻烦你跑一下，最好贴一下mAP啥的

File: docs/zh_cn/advanced_guides/how_to.md [hunk 20]
```diff
@@ -553,26 +553,93 @@ python ./tools/train.py \
...
+
+将上述脚本命名为 `cvt_single_channel.py`, 运行命令为：
 
 ```shell
-CUDA_VISIBLE_DEVICES=0,1,2,3 ./tools/dist_train.sh ${CONFIG} ${GPU_NUM}
+python cvt_single_channel.py  --path data/cat
 ```
 
-这里 `GPU_NUM` 为 4。另外如果在一台机器上多个任务同时多卡训练，需要设置不同的端口，比如以下命令：
+### 新建配置文件并训练
```
Comment (hhaAndroid) [inline docs/zh_cn/advanced_guides/how_to.md:613]: 似乎不需要

File: docs/zh_cn/advanced_guides/how_to.md [hunk 21]
```diff
@@ -553,26 +553,125 @@ python ./tools/train.py \
...
+**目前MMYOLO的一些图像处理函数还不兼容单通道图片，出现不兼容问题概率较高**。为了避免这种不兼容的问题，推荐的做法是将单通道图片作为三通道图片读取后进行训练。这样会降低一些运算性能，但是基本不需要修改配置即可使用。
+
+#### 修改配置文件
+
+以 `projects/misc/custom_dataset/yolov5_s-v61_syncbn_fast_1xb32-100e_cat.py`为 `base` 配置,将其复制到`configs/yolov5`目录下，在同级配置路径下新增 `yolov5_s-v61_syncbn_fast_1xb32-100e_cat_single_channel.py` 文件。
+
+```python
+_base_ = 'yolov5_s-v61_syncbn_fast_1xb32-100e_cat.py'
+
+load_from = './checkpoints/yolov5_s-v61_syncbn_fast_8xb16-300e_coco_20220918_084700-86e02187_single_channel.pth'
```
[the remark this answers is not in the document]
Comment (hhaAndroid) [reply docs/zh_cn/advanced_guides/how_to.md:624]: 这里应该还是先用标准的 coco 权重，然后在下一节说明经过脚本运算后的新配置写法

File: docs/zh_cn/common_usage/single_multi_channel_applications.md.md [hunk 22]
```diff
@@ -0,0 +1,188 @@
...
+
+1. 所有的图片处理 pipeline 都要支持单通道运算
+2. 模型的骨干网络的第一个卷积层输入通道需要从 3 改成 1
+3. 如果希望加载 COCO 预训练权重，则需要处理第一个卷积层权重尺寸不匹配问题
+
+下面以 `cat` 数据集为例，描述整个修改过程，如果你使用的是自定义灰度图像数据集，你可以跳过数据集预处理这一步。
+
+### 1 数据集预处理
+
+自定义数据集的处理训练可参照[自定义数据集 标注+训练+测试+部署 全流程](../user_guides/custom_dataset.md)。
```
Comment (hhaAndroid) [inline docs/zh_cn/common_usage/single_multi_channel_applications.md.md:15]:
```suggestion
自定义数据集的处理训练可参照[自定义数据集 标注+训练+测试+部署 全流程](../recommended_topics/labeling_to_deployment_tutorials.md)。
```

File: docs/en/common_usage/single_multi_channel_applications.md.md [hunk 23]
```diff
@@ -0,0 +1,188 @@
...
+
+1. All image processing pipelines have to support single channel operations
+2. The input channel of the first convolutional layer of the backbone network of the model needs to be changed from 3 to 1
+3. If you wish to load COCO pre-training weights, you need to handle the first convolutional layer weight size mismatch
+
+The following uses the `cat` dataset as an example to describe the entire modification process, if you are using a custom grayscale image dataset, you can skip the dataset preprocessing step.
+
+### 1 Dataset pre-processing
+
+The processing training of the custom dataset can be found in [Annotation-to-deployment workflow for custom dataset](../user_guides/custom_dataset.md)。
```
Comment (hhaAndroid) [inline docs/en/common_usage/single_multi_channel_applications.md.md:15]:
```suggestion
The processing training of the custom dataset can be found in [Annotation-to-deployment workflow for custom dataset](../recommended_topics/labeling_to_deployment_tutorials.md)。
```

**Review by hhaAndroid — APPROVED** [review]

File: docs/zh_cn/common_usage/single_multi_channel_applications.md [hunk 24]
```diff
@@ -0,0 +1,188 @@
+#单通道和多通道应用案例
```
Comment (hhaAndroid) [inline docs/zh_cn/common_usage/single_multi_channel_applications.md:1]:
```suggestion
# 单通道和多通道应用案例
```
